Make fragment-links true JSON

takenspc opened the pull request:
I would like to use fragment-links data to normalize specification URLs. For this purpose, JSON is more suitable format than JavaScript.
Once fragment-links become a JSON, others can more easily reuse this data for their own purposes.

Please note this PR depends on a PR for whatwg/html

sideshowbarker commented:
I’m happy to take a look at this once the https://github.com/whatwg/html/pull/1063 side is ready.

sideshowbarker commented:
Unassigning this from myself for now. But happy to have it re-assigned to me once we have something that we are clear we are ready to consider landing. (Pending https://github.com/whatwg/html/pull/1063 and https://github.com/whatwg/html-build/issues/88)

sideshowbarker commented on src/wattsi.pas:
```diff
@@ -1561,23 +1561,24 @@ function Split(const Document: TDocument; var BigTOC: TElement; const Base: Ansi
-      Writeln(F, '}');
-      Close(F);
+      // link-fixup.js
+      LinkFixup := ReadTextFile('html/link-fixup.js').AsString;
```
I think we can’t assume thereÆs an `html` subdir with a `link-fixup.js` file.

We could instead use `GetEnvironmentVariable` to make wattsi read the value of `HTML_SOURCE` (directory for the clone of the `whatwg/html` repo) and get the `link-fixup.js` file from there—but `HTML_SOURCE` is not a wattsi thing, it’s a `whatwg/html-build` thing, so I think we should also not assume `HTML_SOURCE` is set.

Given all that I think the safest most-portable/robust assumption to build into wattsi is that there’s a `link-fixup.js` in the same directory as whatever file you feed to wattsi as the first parameter from the command line. That will the true if you run wattsi directly in a `whatwg/html` clone (which is something I think we should continue to support), so I think we should make `whatwg/html-build/build.sh` also ensure it’s true (by copying the `link-fixup.js` file from the `HTML_SOURCE` dir to the `HTML_TEMP` dir).

If we agree about all that, then please update the patch in the PR to do something like this:

```diff
diff --git a/src/wattsi.pas b/src/wattsi.pas
index ee51a1e..72c6519 100644
--- a/src/wattsi.pas
+++ b/src/wattsi.pas
@@ -1557,7 +1557,7 @@ begin
    Close(F);
 end;
 
-function Split(const Document: TDocument; var BigTOC: TElement; const Base: AnsiString): Boolean; // True = success, False = failed
+function Split(const Document: TDocument; var BigTOC: TElement; const Base: AnsiString; var LinkFixupJS: AnsiString): Boolean; // True = success, False = failed
 
    procedure SaveFragmentLinks(const Data: Rope);
    var
@@ -1565,7 +1565,7 @@ function Split(const Document: TDocument; var BigTOC: TElement; const Base: Ansi
       LinkFixup: String;
    begin
       // link-fixup.js
-      LinkFixup := ReadTextFile('html/link-fixup.js').AsString;
+      LinkFixup := ReadTextFile(LinkFixupJS).AsString;
       LinkFixup := StringReplace(LinkFixup, '/* WATTSI_INSERTS_FRAGMENT_LINKS_HERE */', Data.AsString, []);
       Assign(LinkFixupFile, Base + '../link-fixup.js');
       Rewrite(LinkFixupFile);
@@ -2209,6 +2209,7 @@ var
    ParamOffset: Integer = 0;
    Source: TFileData;
    OutputDirectory: AnsiString;
+   LinkFixupJS: AnsiString;
    Parser: THTMLParser;
    BigTOC: TElement;
    Documents: array[TVariants] of TDocument;
@@ -2230,6 +2231,12 @@ begin
          exit;
       end;
    OutputDirectory := ParamStr(2 + ParamOffset);
+   LinkFixupJS := ExtractFilePath(ParamStr(1 + ParamOffset)) + 'link-fixup.js';
+   if (not FileExists(LinkFixupJS)) then
+   begin
+      Writeln(LinkFixupJS + ' file not found.');
+      exit;
+   end;
    if (not IsEmptyDirectory(OutputDirectory)) then
    begin
       // only act if, when we start, the output directory is empty, to make sure that the
@@ -2311,7 +2318,7 @@ begin
                   {$IFDEF TIMINGS} Writeln('Splitting spec...'); {$ENDIF}
                   {$IFDEF TIMINGS} StartTime := Now(); {$ENDIF}
                   MkDir(OutputDirectory + '/multipage-' + kSuffixes[Variant]);
-                  if (not Split(Documents[Variant], BigTOC, OutputDirectory + '/multipage-' + kSuffixes[Variant] + '/')) then
+                  if (not Split(Documents[Variant], BigTOC, OutputDirectory + '/multipage-' + kSuffixes[Variant] + '/', LinkFixupJS)) then
                      raise EAbort.Create('Could not split specification');
                   {$IFDEF TIMINGS} Writeln('Elapsed time: ', MillisecondsBetween(StartTime, Now()), 'ms'); {$ENDIF}
                end;
```

takenspc commented:
@sideshowbarker 

Thank you for your review. I addressed your comments and tested with html-build's "1) Use an existing clone on my local filesystem." option.

sideshowbarker commented:
LGTM thanks

Will wait to hear if @domenic agrees

sideshowbarker commented:
I realize now the other alternative here is to add an additional command-line argument for specifying the location of the `link-fixup.js` file—in the same way wattsi currently has command-line arguments for specifying the location of the `bugs.csv` and `caniuse.json` files.

That might be better than assuming the `link-fixup.js` file is always in the same directory as the `source` file.

takenspc commented:
@sideshowbarker 

I added command-line argument to specify path to link-fixup.js
Please refer to this PR and https://github.com/whatwg/html-build/pull/88

domenic commented:
OK, going to start merging these things! We'll see how it goes.
